### PR TITLE
Adjust heatmap / clustergram

### LINKED ISSUE #217: Fix Hierarchical Clustering
## Issue description
Hierarchical Clustering has some issues. Will update this Issue's TODOs as I identify and fix them. Closely related to #201 and I'm working directly on `test-branch-for-paper` for this

## The Problems
- [ ] Saving and reloading a run with Hierarchical Agglomerative Clustering (HAC) doenst work as the `model` data is saved as `!!python/object:sklearn.cluster._agglomerative.AgglomerativeClustering` and there is no constructor to reconstruct the data from that
- [ ] HAC does not produce any plots. Getting the HAC step output from sklearn converted to something that the existing Clustergram plot class supports may be a major pain to implement. Sklearn itself only supports matplotlib dendrograms as outputs
- [ ] The "Plot: Clustergram" step and HAC cannot take input from a t-test
- [ ] The "Plot: Clustergram" step and HAC cannot deal with NaNs

### PR DESCRIPTION
**Pull request by jorisfu**
## Description
Addresses #217 

Changes some things with the clustergram step to more closely match the AML paper

## Changes
- Allow for custom colour scales (low and high colours including respective z-score boundaries)
- Asymmetric colour scales will always retain 0 mapping to white
- Allow for custom user-provided labels for the heatmap colour legend
- Colour legend ticks always show minimum z, maximum z, and 0.
- Change default colours to match AML paper

## Still TODO
- Verify major clusters in our branch matching those in the AML paper supplementary data set

## Testing
Create a Clustergram (e.g. with the AML paper data) and experiment with the new settings

## PR checklist
**Development**
- [x] If necessary, I have updated the documentation (README, docstrings, etc.)
- [x] If necessary, I have created / updated tests.
 
**Mergeability**
- [ ] main-branch has been merged into local branch to resolve conflicts
- [ ] The tests and linter have passed AFTER local merge
- [x] The backend code has been formatted with `black`
 
**Code review**
- [x] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

### CODE REVIEW

File: backend/protzilla/data_analysis/plots.py
Comment (Elena-kal) [inline backend/protzilla/data_analysis/plots.py:230]: doc string should be updated

File: backend/protzilla/data_analysis/plots.py [hunk 1]
```diff
@@ -253,7 +259,14 @@ def clustergram_plot(
             row_colors = None
             color_label_dict = None
 
-        # TODO: Would be nice to actually center values at the z-score of 0
+        if use_custom_colour_scale:
+            custom_colour_scale = (
+                (heatmap_low_colour_limit, heatmap_low_colour),
+                (heatmap_high_colour_limit, heatmap_high_colour),
```
Comment (Elena-kal) [inline backend/protzilla/data_analysis/plots.py]: how sure are we that we use british english in our code base? I always thought it was american in the code and british for everything the user reads and sees. 

Image

File: backend/protzilla/utilities/clustergram.py [hunk 2]
```diff
@@ -528,21 +536,44 @@ def figure(self, computed_traces=None):
...
+                high_colour = self._custom_colour_scale[1][1]
+
+            else:
+                zmin = np.min(heat_data)
+                zmax = np.max(heat_data)
+                low_colour = HEATMAP_LOW_COLOUR
+                high_colour = HEATMAP_HIGH_COLOUR
+
+            # Lerp to keep zero centered
+            midpoint = (0 - zmin) / (zmax - zmin)
```
Comment (Elena-kal) [inline backend/protzilla/utilities/clustergram.py]: How sure are we that it is always zmax != zmin?

File: backend/protzilla/utilities/clustergram.py [hunk 3]
```diff
@@ -528,21 +536,44 @@ def figure(self, computed_traces=None):
...
+
+            # If range of values is only positive or only negative, lerp doesn't work.
+            # In this case we reset. Shouldn't happen for most normalised data sets
+            # which are usually plotted with heatmaps
+            if midpoint < 0 or midpoint > 1.0:
+                midpoint = 0.5
```
Comment (Elena-kal) [inline backend/protzilla/utilities/clustergram.py]: Why 0.5?

**Review by Elena-kal — APPROVED** [review]
It works fine. The plot itself does not look too pretty but I think that is simply the nature of clustergrams. I wrote a couple of comments but overall I think the code is good.

File: backend/protzilla/utilities/clustergram.py [hunk 4]
```diff
@@ -528,21 +536,44 @@ def figure(self, computed_traces=None):
             heat_data = self._data
 
             # symmetrize the heatmap about zero, if necessary
-            if self._center_values:
-                heat_data = np.subtract(heat_data, np.mean(heat_data))
+            # if self._center_values:
+            #     heat_data = np.subtract(heat_data, np.mean(heat_data))
```
Comment (3dot141592) [inline backend/protzilla/utilities/clustergram.py]: Question: Should this centering remain removed permanently, or was it only temporary for the new color scheme?

**Review by 3dot141592 — COMMENTED** [review]
Overall looks good to me aside from the one comment. 

One remaining concern: hierarchical agglomerative clustering still does not accept NaN values. How do we want to handle this going forward? document/expect the user to pre-clean, or implement automatic imputation?

File: backend/protzilla/utilities/clustergram.py [hunk 5]
```diff
@@ -23,6 +23,10 @@
 from sklearn.impute import SimpleImputer
 
 
+HEATMAP_LOW_COLOUR = "#0000FF"
+HEATMAP_HIGH_COLOUR = "#FF0000"
```
Comment (hendraet) [inline backend/protzilla/utilities/clustergram.py]: Why not make this dependent on the min/max values of the `RdBu_r` color to make it more consistent with the usual plotly colorscale? (Might also make it slightly less ugly)

File: backend/protzilla/utilities/clustergram.py (hunk 2 above)
Comment (hendraet) [reply backend/protzilla/utilities/clustergram.py]: Would at least change the default values because executing the step as is definitely raises this ZeroDivisionError

Comment (hendraet) [inline backend/protzilla/utilities/clustergram.py]: Would also add a sanity check that ensures that zmin is actually smaller than zmax. It doesn't throw an error, but it also doesn't look like it behaves correctly (e.g. heatmap range is labeled weirdly)

File: backend/protzilla/data_analysis/plots.py [hunk 6]
```diff
@@ -253,7 +259,14 @@ def clustergram_plot(
             row_colors = None
             color_label_dict = None
 
-        # TODO: Would be nice to actually center values at the z-score of 0
+        if use_custom_colour_scale:
+            custom_colour_scale = (
+                (heatmap_low_colour_limit, heatmap_low_colour),
```
Comment (hendraet) [inline backend/protzilla/data_analysis/plots.py]: Could need some sanity checking or at least a proper error message if user enters sth. that is not a valid color

File: backend/protzilla/utilities/clustergram.py [hunk 7]
```diff
@@ -528,21 +536,44 @@ def figure(self, computed_traces=None):
...
+                midpoint = 0.5
 
             heatmap = go.Heatmap(
                 x=tickvals_col,
                 y=tickvals_row,
                 z=heat_data,
-                colorscale=self._color_map,
-                # TODO: This should be based on the text width of the labels,
-                # or at least passable by the user, so they can adjust it
+                colorscale=[[0, low_colour], [midpoint, "white"], [1, high_colour]],
```
Comment (hendraet) [inline backend/protzilla/utilities/clustergram.py]: Nit: setting `midpoint` to white limits the customizability of the plot. I guess it might be a bit too much to also add a field to customize this variable, but since I thought about it, I just wanted to leave this here.

File: backend/protzilla/utilities/clustergram.py [hunk 8]
```diff
@@ -528,21 +536,44 @@ def figure(self, computed_traces=None):
...
+                zmin=zmin,
+                zmax=zmax,
                 colorbar=dict(
-                    title="Heatmap Legend",
+                    title=self._heatmap_legend_title,
                     yanchor="bottom",
                     y=0.0,
                     len=1 / self.colorbar_count,
+                    tickmode="array",
+                    tickvals=(zmin, 0, zmax),
```
Comment (hendraet) [inline backend/protzilla/utilities/clustergram.py]: Midpoint is not shown in legend if both values are positive/negative

**Review by hendraet — CHANGES_REQUESTED** [review]
Works nicely, and I really like the added customization options. 
The only thing I'm missing is better error handling and corresponding test cases, i.e., what happens if zmin and zmax are the same or inverted and what if the user enters an invalid color.

File: backend/protzilla/data_analysis/plots.py (hunk 6 above)
Comment (jorisfu) [reply backend/protzilla/data_analysis/plots.py]: Fixed this by adding a custom colour picker, is cleaner anyway and doesn't require sanity checking down the line (since plotly can also take other values for colours apart from hex codes)

File: backend/protzilla/utilities/clustergram.py (hunk 4 above)
Comment (jorisfu) [reply backend/protzilla/utilities/clustergram.py]: mistake on my part, should not have been deleted.

File: backend/protzilla/utilities/clustergram.py (hunk 5 above)
Comment (jorisfu) [reply backend/protzilla/utilities/clustergram.py]: Fascinatingly, using min/max of RdBu doesn't look that good, because RdBu uses more than min/mid/max colours to interpolate:

mynd

Should I still change it to this? I adjusted it to these defaults to match the (ugly) colour scale in the AML paper

File: backend/protzilla/utilities/clustergram.py (hunk 7 above)
Comment (jorisfu) [reply backend/protzilla/utilities/clustergram.py]: Fair point, this could be addressed in the future. Helps people create more misleading plots as well :)

**Comment by jorisfu** [comment]
@hendraet @Elena-kal I addressed every other comment y'all made with these commits. Since this changes a lot of what I previously did all over again and I also added a new front-end component (colour picker so we don't need to validate colours down the line), it'd be cool if you could re-review

File: backend/protzilla/utilities/clustergram.py (hunk 5 above)
Comment (hendraet) [reply backend/protzilla/utilities/clustergram.py]: I admit, it's a bit underwhelming (although I still like it slightly better than the original colors). But I totally get the notion to adapt it to the color scale in the AML paper, so I wouldn't complain about using these colors (since they are customizable anyways)

File: backend/protzilla/data_analysis/plots.py (hunk 6 above)
Comment (hendraet) [reply backend/protzilla/data_analysis/plots.py]: During review, I thought that a color picker would be ideal, but I deemed it too much work to propose it. I love that you did implement one anyways :blush:

File: backend/protzilla/utilities/clustergram.py [hunk 9]
```diff
@@ -531,18 +540,53 @@ def figure(self, computed_traces=None):
...
+                zmin = np.min(heat_data)
+                zmax = np.max(heat_data)
+                low_color = HEATMAP_LOW_COLOR
+                high_color = HEATMAP_HIGH_COLOR
+
+                # Really only zmin == zmax can happen here
+                if not zmin < zmax:
+                    raise ValueError(
+                        "Data consists only of identical values. Not plotting."
+                    )
```
Comment (hendraet) [inline backend/protzilla/utilities/clustergram.py:564]: Nit: move this check to the top, or at least before the trace calculation. Otherwise, you have to wait for the computation to finish, which might take some time, only to get an error afterwards.

Comment (hendraet) [reply backend/protzilla/utilities/clustergram.py:564]: Maybe same with the color check above (although I'm not sure if that one is needed anymore)

**Review by hendraet — COMMENTED** [review]
I like all the added improvements! Only a minor suggestion

**Review by hendraet — APPROVED** [review]

File: backend/protzilla/utilities/clustergram.py (hunk 9 above)
Comment (jorisfu) [reply backend/protzilla/utilities/clustergram.py:564]: I added the same check in `_compute_clustered_data`, just to make sure if calculation changes somehow we still have this one in place.
